compiler/qsc_qasm3/src/io.rs

```diff
@@ -1,37 +1,46 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
+mod error;
+pub use error::Error;
+pub use error::ErrorKind;
+
 use std::{
     path::{Path, PathBuf},
     sync::Arc,
 };
 
 use rustc_hash::FxHashMap;
 
-use miette::IntoDiagnostic;
-
 /// A trait for resolving include file paths to their contents.
 /// This is used by the parser to resolve `include` directives.
 /// Implementations of this trait can be provided to the parser
 /// to customize how include files are resolved.
 pub trait SourceResolver {
+    fn ctx(&mut self) -> &mut SourceResolverContext;
+
     #[cfg(feature = "fs")]
-    fn resolve<P>(&self, path: P) -> miette::Result<(PathBuf, String)>
+    fn resolve<P>(&mut self, path: P) -> miette::Result<(PathBuf, String), Error>
     where
         P: AsRef<Path>,
     {
         let path = std::fs::canonicalize(path).map_err(|e| {
-            crate::Error(crate::ErrorKind::IO(format!(
+            Error(ErrorKind::IO(format!(
                 "Could not resolve include file path: {e}"
             )))
         })?;
+
+        self.ctx().check_include_errors(&path)?;
+
         match std::fs::read_to_string(&path) {
-            Ok(source) => Ok((path, source)),
-            Err(_) => Err(crate::Error(crate::ErrorKind::NotFound(format!(
+            Ok(source) => {
+                self.ctx().add_path_to_include_graph(path.clone());
+                Ok((path, source))
+            }
+            Err(_) => Err(Error(ErrorKind::NotFound(format!(
```

Comment:
Should this be "not found"? Looks like you'd get here in the case of a read error. In any case, should we be wrapping/propagating the original error message from `read_to_string` here?

Reply:
https://github.com/microsoft/qsharp/pull/2303